compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -427,4 +476,386 @@ LogicalResult splitLogicalObjectFifos(
+  SmallVector<OpFoldResult> sourceOffsetsA = dmaOpA.getSourceMixedOffsets();
+  SmallVector<OpFoldResult> sourceSizesA = dmaOpA.getSourceMixedSizes();
+  SmallVector<OpFoldResult> sourceStridesA = dmaOpA.getSourceMixedStrides();
+  SmallVector<OpFoldResult> sourceOffsetsB = dmaOpB.getSourceMixedOffsets();
+  SmallVector<OpFoldResult> sourceSizesB = dmaOpB.getSourceMixedSizes();
+  SmallVector<OpFoldResult> sourceStridesB = dmaOpB.getSourceMixedStrides();
+
+  SmallVector<OpFoldResult> targetOffsetsA = dmaOpA.getTargetMixedOffsets();
+  SmallVector<OpFoldResult> targetSizesA = dmaOpA.getTargetMixedSizes();
+  SmallVector<OpFoldResult> targetStridesA = dmaOpA.getTargetMixedStrides();
+  SmallVector<OpFoldResult> targetOffsetsB = dmaOpB.getTargetMixedOffsets();
+  SmallVector<OpFoldResult> targetSizesB = dmaOpB.getTargetMixedSizes();
+  SmallVector<OpFoldResult> targetStridesB = dmaOpB.getTargetMixedStrides();
+
+  SmallVector<OpFoldResult> newSourceOffsets = sourceOffsetsA;
+  SmallVector<OpFoldResult> newSourceSizes = sourceSizesA;
+  SmallVector<OpFoldResult> newSourceStrides = sourceStridesA;
+  if (failed(combineL3ToL2AccessPatterns(
+          rewriter, sourceOffsetsA, sourceSizesA, sourceStridesA,
+          sourceOffsetsB, sourceSizesB, sourceStridesB, newSourceOffsets,
+          newSourceSizes, newSourceStrides, splitDims, nonSplitDims))) {
+    return failure();
+  }
+
+  SmallVector<OpFoldResult> newTargetOffsets = targetOffsetsA;
+  SmallVector<OpFoldResult> newTargetSizes = newSourceSizes;
+  SmallVector<OpFoldResult> newTargetStrides = targetStridesA;
+  // Now we need to create a new L2 buffer based on `newTargetSizes`.
+  LogicalObjectFifoFromMemrefOp oldL2ObjectFifo = dmaOpA.getTargetObjectFifo();
+  AMDAIE::LogicalObjectFifoFromMemrefOp newL2ObjectFifo =
+      createNewLogicalObjectFifo(rewriter, oldL2ObjectFifo, newTargetSizes);
+
+  // Create combined L3->L2 Dma.
+  rewriter.setInsertionPoint(dmaOpA);
+  auto combinedL3ToL2DmaOp = rewriter.create<AMDAIE::DmaCpyNdOp>(
+      dmaOpA.getLoc(), newL2ObjectFifo, llvm::ArrayRef(newTargetOffsets),
+      llvm::ArrayRef(newTargetSizes), llvm::ArrayRef(newTargetStrides),
+      dmaOpA.getSource(), llvm::ArrayRef(newSourceOffsets),
+      llvm::ArrayRef(newSourceSizes), llvm::ArrayRef(newSourceStrides));
+  // Replace the uses of 2nd L3->L2 Dma with the new combined L3->L2 Dma
+  // and erase the 1st L3->L2 Dma.
+  rewriter.replaceOp(dmaOpB, combinedL3ToL2DmaOp);
+  rewriter.eraseOp(dmaOpA);
+  return newL2ObjectFifo;
+}
+
+/// Utility to fetch a unique CoreOp associated with a L2->L1 Dma op.
+static CoreOp fetchUniqueCoreOp(DmaCpyNdOp &l2ToL1DmaOp) {
+  SmallVector<CoreOp> coreOps;
+  for (Operation *userOp : l2ToL1DmaOp->getUsers()) {
+    if (auto coreOp = dyn_cast<CoreOp>(userOp)) {
+      coreOps.push_back(coreOp);
+    }
+  }
+  assert(coreOps.size() == 1 &&
+         "L2->L1 Dma op expected to have a unique Core op");
+  return coreOps[0];
+}
```

Comment:
Return an `std::optional<CoreOp>` instead? Also, this could be moved to the DmaUtils?